_posts/2021-01-28-migracja-schematów-db.md

```diff
@@ -0,0 +1,102 @@
+  które można byłoby zaoszczędzić gdyby migracja A była wykonana osobno. 
+* Wykonywane zmiany powinny być przyrostowe, czyli zmiana dla danej wersji powinna być uruchomiona tylko raz.
+
+#### Przykładowe biblioteki
+* Liquibase
+  * [liquibase.org](https://www.liquibase.org/)
+* FlywayDB
+  * [flywaydb.org](https://flywaydb.org/)
+* MyBatis Migration
+  * [mybatis.org/migrations](https://mybatis.org/migrations)
+* w przypadku bazy MongoDB
+  * Mongock
+    * [github.com/cloudyrock/mongock](https://github.com/cloudyrock/mongock)
+    * [mongock.io](https://www.mongock.io)
+  * Migrate-mongo
+    * [npmjs.com/package/migrate-mongo](https://www.npmjs.com/package/migrate-mongo)
+    * [github.com/seppevs/migrate-mongo](https://github.com/seppevs/migrate-mongo)
+  * liquibase-mongodb
+    * [github.com/liquibase/liquibase-mongodb](https://github.com/liquibase/liquibase-mongodb)
+
+w powyższych bibliotekach tworzenie i uruchamianie przyrostowe zmian jest wbudowane.
+
+### Zmiany powinny być kompatybilne wstecz
+Jest to szczególnie ważne jeśli chcemy bezprzerwowo aktualizować naszą bazę danych, w takim wypadku aktualizacja bazy danych nie powinna spowodować że starsza wersja aplikacji przestanie działać.
+
+Przykładowo nowa kolumna powinna mieć domyślną wartość lub przyjmować null-e.
+
+Zmiana nazwy kolumny lub jej usunięcie powinno być rozbite na kilka etapów, tak aby jej prawdziwe usunięcie było wykonane nie w docelowej wersji tylko np. w następnej iteracji, jak będziemy pewni że żadna aplikacja z niej nie korzysta.
+
+## Aplikowanie zmian
+Aplikowania zmian wykonanych w ramach ewolucyjnej bazy danych, jest już zależne od konkretnego przypadku.
+
+Jeśli baza danych jest ściśle związana jedną z aplikacją, możemy ją uruchamiać bezpośrednio z kodu [🔗⁵](https://andrewlock.net/deploying-asp-net-core-applications-to-kubernetes-part-7-running-database-migrations/#running-migrations-on-application-startup) .
+
+W przypadku gdy aplikacja jest rozproszona i nie chcemy blokować wszystkich instancji aplikacji na czas migracji, lub kilka różnych aplikacji korzysta z tej bazy danych, możemy uruchamiać migrację niezależnie od aplikacji.
+
+1. Wykonywanie zmian uruchamianych za pomocą CI/CD (np. automatycznie po otrzymaniu nowej wersji)
+   wykonujemy merge z migracjami do master → Jenkins wykrywa zmianę na repo i uruchamia migrację na bazie wskazanej w konfiguracji.
+2. Z wykorzystaniem mechanizmów dostarczonych przez platformę, na której będzie to uruchamiane
+   * w Kubernetes
+     * wykorzystanie initContainers aby odpalić migrację przed uruchomieniem docelowego kontenera z aplikacją (w takim wypadku każda replika odpali migrację, a to mechanizm migracji musi zapewnić lock oraz nie wykonanie ponownie tych samych migracji) [🔗⁵](https://andrewlock.net/deploying-asp-net-core-applications-to-kubernetes-part-7-running-database-migrations/#init-containers) ,
```

Comment:
tutaj chętnie poznam opinię kogoś na drugą rękę, coś mi tutaj tak melodyjnie się nie składa pod kątem układu. Tak na szybko, to taka propozycja, ale jak ktoś jeszcze uzna, że to jest zrozumiałe, to nie mam wniosku poza poszerzeniem o "migrację schematu bazy danych" z powodu SEO.
==> wykorzystanie initContainers, celem odpalenia migracji schematu bazy danych przed uruchomieniem docelowego kontenera z aplikacją (w takim wypadku każda replika odpali migrację, a to właśnie mechanizm migracji musi zapewnić lock oraz nie wykonanie ponownie tych samych migracji)

Reply:
poprawiłem na:

```
wykorzystanie initContainers, celem odpalenia migracji schematu bazy danych przed uruchomieniem docelowego kontenera z aplikacją 
       (w takim wypadku każda replika odpali migrację schematu, a to mechanizm migracji musi zapewnić, że zmiany zostaną wykonane tylko raz i wszystkie na jednym kontenerze)
```